toolchain/sem_ir/inst_kind.h

```diff
@@ -16,14 +16,37 @@ namespace Carbon::SemIR {
 // of value.
 enum class InstValueKind : int8_t {
   // This instruction doesn't produce a value, and shouldn't be referenced by
-  // other
-  // instructions.
+  // other instructions.
   None,
   // This instruction represents an expression or expression-like construct that
   // produces a value of the type indicated by its `type_id` field.
   Typed,
 };
 
+// Whether an instruction can be used to define a constant value. This specifies
+// whether the instruction can be added to the `constants()` list. Note that
+// even instructions that cannot define a constant value can still have an
+// associated `constant_value()`, but the constant value will be a different
+// kind of instruction.
+enum class InstConstantKind : int8_t {
+  // This instruction never defines a constant value. For example,
+  // `UnaryOperatorNot` never defines a constant value; if its operand is a
+  // template constant, its constant value will instead be a `BoolLiteral`. This
+  // is also used for instructions that don't produce a value at all.
+  Never,
+  // This instruction may be a symbolic constant, depending on its operands, but
+  // is never a template constant. For example, a `Call` instruction can be a
+  // symbolic constant but never a template constant.
```

Comment:
Is this going to remain true with metaprogramming? I'm kind of wondering if BindSymbolicName might be a better example, although it's a little more tautological.

Reply:
I would expect this to remain true -- note that we're classifying the kind of instructions that define constants, not merely the ones that can evaluate to constants, so if we have a call that produces a constant value, the corresponding instruction kind will be the one corresponding to the result of the call, not the `Call` instruction itself. (For example, evaluating a `Call` instruction to a function that returns 42 will produce a constant of kind `IntLiteral`, not of kind `Call`.) I picked this example to draw attention to that subtlety.